src/jupyterlabcontroller/services/lab.py

```diff
@@ -69,175 +206,339 @@ async def create_lab(
         # Check to see if the lab already exists. If so, but it is in a failed
-        # state, delete it first.
-        status = await self._lab_state.get_lab_status(user.username)
-        delete_first = status == LabStatus.FAILED
+        # state, we will delete the previous lab first.
+        state = await self.get_lab_state(user.username)
+        delete_first = bool(state and state.status == LabStatus.FAILED)
 
-        # Start the spawning process. This also checks for conflicts and
-        # raises an exception if the lab already exists and is not in a failed
-        # state.
+        # If there is any operation already in progress, raise an error.
+        # Similarly, if the lab already exists and is not in a failed state,
+        # raise an error.
         #
-        # A LabManager is per-request, so the management of the background
-        # task that does the lab spawning (which outlasts the request that
-        # kicks it off) is handed off to LabStateManager here.
-        resources = self._size_manager.resources(lab.options.size)
-        await self._lab_state.start_lab(
-            username=user.username,
-            state=UserLabState.from_request(user, lab, resources),
-            spawner=partial(
-                self._spawn_lab, user, lab, image, delete_first=delete_first
-            ),
-            start_progress=35,
-            end_progress=75,
+        # This must be done after any other preliminaries that may yield
+        # control, to ensure that state doesn't change between our check and
+        # starting a new operation.
+        lab = self._labs[username]
+        if lab.monitor.in_progress:
+            operation_type = lab.monitor.in_progress.value
+            msg = f"Operation in progress for {username}: {operation_type}"
+            raise OperationConflictError(msg)
+        if lab.state and not delete_first:
+            raise LabExistsError(f"Lab already exists for {username}")
+
+        # Kick off the spawn and hand it off to the monitor to watch.
+        #
+        # Since we may yield control when start_spawn is called but before it
+        # can run, store the new lab state *after* the monitor has
+        # successfully started. If the user has no lab and two spawns are
+        # started at the same time, the second will raise an
+        # OperationConflictError. This ordering ensures that only the
+        # successful call is able to update the user's lab state.
+        lab.events.clear()
+        resources = self._size_manager.resources(spec.options.size)
+        state = UserLabState.from_request(user, spec, resources)
+        spawner = self._spawn_lab(
+            user=user,
+            state=state,
+            spec=spec,
+            image=image,
+            events=lab.events,
+            delete_first=delete_first,
         )
+        operation = _Operation(_LabOperation.SPAWN, spawner, state, lab.events)
+        await lab.monitor.monitor(operation, self._spawner_done)
```

Comment:
I think I see: we can't just protect the spawn with a mutex because we want multiple people to spawn at the same time, but a given person has to spawn exactly once.

Reply:
Right, so we need per-user mutexes, and it seemed cleaner when I put those in separate classes instead of trying to juggle everything in the LabManager class.